### PR TITLE
PARQUET-1826: Document Hadoop configuration options

### PR DESCRIPTION
**Pull request by garawalid**
Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Parquet Jira](https://issues.apache.org/jira/browse/PARQUET/) issues and references them in the PR title. For example, "PARQUET-1234: My Parquet PR"
  - https://issues.apache.org/jira/browse/PARQUET-XXX
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

**Comment by garawalid** [comment]
@gszadovszky could you please take a look if you have some free time?  
Also, I couldn't find a good description of `parquet.read.filter
` and `parquet.private.read.filter.predicate`.

**Comment by gszadovszky** [comment]
Both keys are for setting the filter for the predicate push down.
`parquet.read.filter` is for the old filter API can be found in the package [org.apache.parquet.filter](https://github.com/apache/parquet-mr/tree/master/parquet-column/src/main/java/org/apache/parquet/filter). One may specify the class name of the implementation of an [UnboundRecordFilter](https://github.com/apache/parquet-mr/blob/master/parquet-column/src/main/java/org/apache/parquet/filter/UnboundRecordFilter.java). See details at [ParquetInputFormat.getUnboundRecordFilterInstance](https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetInputFormat.java#L191-L207).
`parquet.private.read.filter.predicate` is for the new filter API can be found in the package [org.apache.parquet.filter2](https://github.com/apache/parquet-mr/tree/master/parquet-column/src/main/java/org/apache/parquet/filter2). The value is a gzip compressed base64 encoded java serialized object so it is not for writing directly to the configuration. I guess, that's why it is called `private`. One may use [ParquetInputFormat.setFilterPredicate](https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetInputFormat.java#L217-L227) to set it. There is also a key `parquet.private.read.filter.predicate.human.readable` where a human readable form of the predicate is saved but it is never read from the code. It is only for the user who might be interested in the configuration values.
The user may use either the old or the new style of the filter (suggested to use the new one) but not both of them.

File: parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetOutputFormat.java [hunk 1]
```diff
@@ -90,7 +90,7 @@
  *
  * # Maximum size (in bytes) allowed as padding to align row groups
  * # This is also the minimum size of a row group. Default: 0
- * parquet.writer.max-padding=2097152 # 2 MB
+ * parquet.writer.max-padding=8388608 # 8 MB
```
[the remark this answers is not in the document]
Comment (garawalid) [reply parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetOutputFormat.java:93]: The default value equals to 8MB here: https://github.com/apache/parquet-mr/blob/d00b2f105f9f732e310ed43c7bfb318213e1ac81/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetWriter.java#L55

I should also change the comment above to `Default: 8388608`

File: parquet-hadoop/README.md [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+
+Job writeJob = new Job(conf);
+ParquetOutputFormat.setBlockSize(writeJob, 1024);
+
+```
+
+## Class: ParquetOutputFormat
+
+**Property:** `parquet.summary.metadata.level`  
+**Description:** Write summary files. If this property is set to `all`, write both summary file with row group info (_metadata) and summary file without (_common_metadata).
```
[the remark this answers is not in the document]
Comment (garawalid) [reply parquet-hadoop/README.md:46]: The file name generated within the same directory as parquet files. I'll clarify it

File: parquet-hadoop/README.md [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+  
+# Hadoop integration
+
+**Todo:** Add a description
+
+## Properties
+
+It's possible to configure the [ParquetInputFormat](https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetInputFormat.java) / [ParquetOutputFormat](https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetOutputFormat.java) with Hadoop config or programmatically with setters.
+
+**Note:** Properties passed through InputOutputParquetFormat have a high precedence than properties passed via Hadoop Configuration.
```
[the remark this answers is not in the document]
Comment (garawalid) [reply parquet-hadoop/README.md:28]: You are right! I'll delete the

File: parquet-hadoop/README.md [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+**Property:** `parquet.enable.summary-metadata`  
+**Description:** This property is deprecated, use `parquet.summary.metadata.level` instead.  
+If it is `true`, it similar to `parquet.summary.metadata.level` with `all`. If it is `false`, it is similar to `NONE`.  
+**Default value:** true  
+
+---
+
+**Property:** `parquet.block.size`  
+**Description:** The block size in bytes is the size of a row group being buffered in memory. This limits the memory usage when writing.  
+Larger values will improve the IO when reading but consume more memory when writing.
```
[the remark this answers is not in the document]
Comment (garawalid) [reply parquet-hadoop/README.md:62]: Thanks @gszadovszky for the explanation. What about this description? 

The block size in bytes. This property depends on the file system: 
- If the file system (FS) used supports blocks like HDFS, the block size will be the maximum between the default block size of FS and this property. And the row group size will be equals to this property.
block_size = max(default_fs_block_size, `parquet.block.size`) 
row_group_size = parquet.block.size`

- If the file system used doesn't support blocks, then this property will define the row group size.

Note that larger values of row group size will improve the IO when reading but consume more memory when writing

File: parquet-hadoop/README.md [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+conf.set("parquet.bloom.filter.enabled#column.path", false); // Disable the bloom filter for the column 'column.path'
+// The bloom filter will be enabled for all columns except 'column.path'
+```
+
+---
+
+**Property:** `parquet.bloom.filter.expected.ndv`  
+**Description:** The expected number of distinct values in a column, it is used to compute the optimal size of the bloom filter.  
+Note that if this property is not enabled, the bloom filter will use the maximum size.  
+If this is property is set without a the column name, all columns will have a boom filter with the same expected number of distinct values.
```
[the remark this answers is not in the document]
Comment (garawalid) [reply parquet-hadoop/README.md:200]: Ok! I'll update it

File: parquet-hadoop/README.md [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+**Description:** This property is deprecated, use `parquet.summary.metadata.level` instead.  
+If it is `true`, it similar to `parquet.summary.metadata.level` with `all`. If it is `false`, it is similar to `NONE`.  
+**Default value:** true  
+
+---
+
+**Property:** `parquet.block.size`  
+**Description:** The block size in bytes is the size of a row group being buffered in memory. This limits the memory usage when writing.  
+Larger values will improve the IO when reading but consume more memory when writing.  
+**Default value:** 134217728 (128 MB)
```
[the remark this answers is not in the document]
Comment (garawalid) [reply parquet-hadoop/README.md]: I agree!

File: parquet-hadoop/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (gszadovszky) [reply parquet-hadoop/README.md]: perfect :)

File: parquet-hadoop/README.md [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+
+**Property:** `parquet.page.size.row.check.min`  
+**Description:** The minimum number of row per page.  
+**Default value:** 100
+
+---
+
+**Property:** `parquet.page.size.row.check.max`  
+**Description:** The maximum number of row per page.  
+**Default value:** 10000
```
[the remark this answers is not in the document]
Comment (garawalid) [reply parquet-hadoop/README.md]: Thanks for the clarification, now it makes sense :)

**Comment by garawalid** [comment]
Thanks for your help @gszadovszky. I addressed your comments.

File: parquet-hadoop/README.md [hunk 8]
```diff
@@ -58,9 +58,15 @@ If it is `true`, it similar to `parquet.summary.metadata.level` with `all`. If i
 ---
 
 **Property:** `parquet.block.size`  
-**Description:** The block size in bytes.
-is the size of a row group being buffered in memory. This limits the memory usage when writing.  
-Larger values will improve the IO when reading but consume more memory when writing.  
+**Description:** The block size in bytes. This property depends on the file system:
+
+- If the file system (FS) used supports blocks like HDFS, the block size will be the maximum between the default block size of FS and this property. And the row group size will be equals to this property.
```
Comment (gszadovszky) [inline parquet-hadoop/README.md:63]: nit:
```suggestion
- If the file system (FS) used supports blocks like HDFS, the block size will be the maximum between the default block size of FS and this property. And the row group size will be equal to this property.  
```

File: parquet-hadoop/README.md [hunk 9]
```diff
@@ -158,7 +164,7 @@ This property should be between 0 and 1.
 ---
 
 **Property:** `parquet.page.size.row.check.max`  
-**Description:** The maximum number of rows per page.  
+**Description:** The frequency of checks of the page size limit. In other words, we perform the checking after each `parquet.page.size.row.check.max` rows.
```
Comment (gszadovszky) [inline parquet-hadoop/README.md:167]: It is not that simple and `parquet.page.size.row.check.min` shall be updated as well.
So, the frequency of the checks will be between `parquet.page.size.row.check.min` and `parquet.page.size.row.check.max`. You may influence the checking frequency by setting these numbers. (Higher frequency results more precise page sizes, lower frequency improves performance.) See the [code](https://github.com/apache/parquet-mr/blob/master/parquet-column/src/main/java/org/apache/parquet/column/impl/ColumnWriteStoreBase.java#L221-L266) for more details.

File: parquet-hadoop/README.md [hunk 10]
```diff
@@ -230,23 +236,28 @@ conf.set("parquet.bloom.filter.expected.ndv#column.path", 200)
...
 
-**Property:** `parquet.strict.typing`  
-**Description:** Whether to enable type checking for conflicting schema.  
-**Default value:** `true`
+ **Property:** `parquet.private.read.filter.predicate`  
+ **Description:** The filter class used in the new filter API in the package `org.apache.parquet.filter2.predicate`
+ Note that this class should implements `org.apache.parquet.filter2..FilterPredicate` and the value of this property should be a gzip compressed base64 encoded java serialized object.  
+ The new filter API can filter records or filter entire row groups of records without reading them at all.
+
+**Note:** User should either use the old filter API (`parquet.read.filter`) or the new one (`parquet.private.read.filter.predicate`).
```
Comment (gszadovszky) [inline parquet-hadoop/README.md:254]: We shall suggest using the new one.

File: parquet-hadoop/README.md [hunk 11]
```diff
@@ -230,23 +236,28 @@ conf.set("parquet.bloom.filter.expected.ndv#column.path", 200)
...
+**Description:** The filter class name that implements `org.apache.parquet.filter.UnboundRecordFilter`. This class is for the old filter API in the package `org.apache.parquet.filter`, it filters records during record assembly.
 
 ---
 
-**Property:** `parquet.strict.typing`  
-**Description:** Whether to enable type checking for conflicting schema.  
-**Default value:** `true`
+ **Property:** `parquet.private.read.filter.predicate`  
+ **Description:** The filter class used in the new filter API in the package `org.apache.parquet.filter2.predicate`
+ Note that this class should implements `org.apache.parquet.filter2..FilterPredicate` and the value of this property should be a gzip compressed base64 encoded java serialized object.
```
Comment (gszadovszky) [inline parquet-hadoop/README.md:251]: It is not a class but a serialized java object structure. But I am not sure if we would like to describe it in details. What is more interesting to the user is that this value shall not be set _manually_, the method [ParquetInputFormat.setFilterPredicate](https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetInputFormat.java#L217-L227) shall be used instead.
We might also mention the property `parquet.private.read.filter.predicate.human.readable` that contains the filter in a human readable format for the users to help understanding the configuration. (The human readable format is not used for anything else.)

File: parquet-hadoop/README.md (hunk 10 above)
Comment (garawalid) [reply parquet-hadoop/README.md:254]: I agree!

File: parquet-hadoop/README.md (hunk 9 above)
Comment (garawalid) [reply parquet-hadoop/README.md:167]: Thanks @gszadovszky for the clarification, I'll duplicate the description for both `parquet.page.size.row.check.min` and `parquet.page.size.row.check.max` properties.

File: parquet-hadoop/README.md (hunk 11 above)
Comment (garawalid) [reply parquet-hadoop/README.md:251]: I think it's okay if we keep the details of the object. After all, we will suggest using the `setFilterPredicate` method.

**Comment by garawalid** [comment]
Thanks @gszadovszky for reviewing this PR.

**Comment by gszadovszky** [comment]
Tanks, @garawalid for working on it! :)

**Pull request merged**